httpc.go

```diff
@@ -303,60 +303,68 @@ var (
 	errorEncodedRawBodyCollision = errors.New("cannot use both body encoding and raw body content")
 )
 
-func (r *Request) prepBody() (body io.ReadCloser, err error) {
+func (r *Request) setBody(req *http.Request) (err error) {
 
 	// If requested, parse the requst body using the specified encoder
+	bodyBytes := r.body
 	if r.bodyEncoder != nil {
 		if len(r.body) > 0 {
-			return body, errorEncodedRawBodyCollision
+			return errorEncodedRawBodyCollision
 		}
 
-		r.body, err = r.bodyEncoder.Encode()
+		bodyBytes, err = r.bodyEncoder.Encode()
 		if err != nil {
-			return body, fmt.Errorf("error encoding body: %w", err)
+			return fmt.Errorf("error encoding body: %w", err)
 		}
+		req.Header.Set("Content-Type", r.bodyEncoder.ContentType())
 	}
 
-	if len(r.body) == 0 {
-		return http.NoBody, nil
-	}
+	contentLength := len(bodyBytes)
+
+	// From net/http:
+	//
+	// 	If body is of type *bytes.Buffer, *bytes.Reader, or *strings.Reader, the returned request's ContentLength is set to its exact value
+	// 	(instead of -1), GetBody is populated (so 307 and 308 redirects can replay the body), and Body is set to NoBody if the ContentLength
+	// 	is 0.
+	req.Body = http.NoBody
+	if contentLength > 0 {
+		req.ContentLength = int64(contentLength)
+
+		// If a delay was requested, assign a delayed reader
+		if r.delay > 0 {
+			dr := newDelayedReader(bytes.NewBuffer(bodyBytes), r.delay)
 
-	// If a delay was requested, assign a delayed reader
-	if r.delay > 0 {
-		return io.NopCloser(newDelayedReader(bytes.NewBuffer(r.body), r.delay)), nil
+			req.GetBody = func() (io.ReadCloser, error) {
+				snapshot := newDelayedReader(bytes.NewReader(bodyBytes), r.delay)
+				return io.NopCloser(snapshot), nil
+			}
+			req.Body = io.NopCloser(dr)
+			return nil
+		}
+
+		buf := bytes.NewReader(bodyBytes)
+
+		req.GetBody = func() (io.ReadCloser, error) {
+			return io.NopCloser(bytes.NewReader(bodyBytes)), nil
```

Comment:
Isn't this a duplicate of the just-allocated `buf := bytes.NewReader(bodyBytes)` (maybe see also my comment below about potentially re-using the body in a seekable buffer)?